src/pip/_internal/index.py

```diff
@@ -626,32 +674,11 @@ def find_requirement(self, req, upgrade):
         Raises DistributionNotFound or BestVersionAlreadyInstalled otherwise
         """
         all_candidates = self.find_all_candidates(req.name)
-
-        # Filter out anything which doesn't match our specifier
-        compatible_versions = set(
-            req.specifier.filter(
-                # We turn the version object into a str here because otherwise
-                # when we're debundled but setuptools isn't, Python will see
-                # packaging.version.Version and
-                # pkg_resources._vendor.packaging.version.Version as different
-                # types. This way we'll use a str as a common data interchange
-                # format. If we stop using the pkg_resources provided specifier
-                # and start using our own, we can drop the cast to str().
-                [str(c.version) for c in all_candidates],
-                prereleases=(
-                    self.allow_all_prereleases
-                    if self.allow_all_prereleases else None
-                ),
-            )
-        )
-        applicable_candidates = [
-            # Again, converting to str to deal with debundling.
-            c for c in all_candidates if str(c.version) in compatible_versions
-        ]
-
+        applicable_candidates = list(self._filter_applicable_candidates(
+            all_candidates, specifier=req.specifier,
+        ))
         if applicable_candidates:
-            best_candidate = max(applicable_candidates,
-                                 key=self._candidate_sort_key)
+            best_candidate = self._select_best_candidate(applicable_candidates)
```

Comment:
I read and appreciate your comments describing your thinking on this. Thanks. While I think I understand why you did it, on balance it seems to me like it would be better to avoid copying the exact three lines right next to each other, even if this is something you plan to fix later. It seems like it would be simpler and clearer and less error-prone to have them both go through the same function.

Two easily implementable alternatives, both of which I'm sure you've already thought of, and either of which I'd be okay with--

1. You could add a `return_intermediate_values=False` argument that `find_requirement()` could pass `True` for. Passing `True` would cause the return value to be a tuple rather than a single value. It's ugly, but would get the job done (and at least it wouldn't affect `pip_version_check()`'s invocation).

2. Kind of like `subprocess.run()`'s API that returns a `CompletedProcess` object, you could return an object instead of a single value. You could even have two variants: `analyze / plan / calculate_best_candidate()` that returns an object containing auxiliary info, and a simpler wrapper function `find_best_candidate()` that `pip_version_check()` could use that returns just the final value. As for (1), the latter would make it so `pip_version_check()` would be unaffected by the extra info.

Reply:
The `CompletedProcess` model sounds like a good idea :D In retrospect it is not unlike the original tuple solution, but with a better interface.